File: ROS2-basic/chapter5/chapter5_ws/src/demo_python_tf/demo_python_tf/static_tf_broadcaster.py
import math
import rclpy
from rclpy.node import Node
from tf2_ros import StaticTransformBroadcaster       # 静态坐标发布器
from geometry_msgs.msg import TransformStamped       # 消息接口
from tf_transformations import quaternion_from_euler # 欧拉角转四元数函数

class StaticTFBroadcaster(Node):
    def __init__(self):
        super().__init__('static_tf_broadcaster')
        self.static_broadcaster_ = StaticTransformBroadcaster(self)
        self.publish_static_tf()

    def publish_static_tf(self):
        transform = TransformStamped()
        transform.header.stamp = self.get_clock().now().to_msg()
        transform.header.frame_id = "base_link"
        transform.child_frame_id = "camera_link"

        # 设置平移部分
        transform.transform.translation.x = 0.5
        transform.transform.translation.y = 0.5
        transform.transform.translation.z = 0.5
 
        # 欧拉角转四元数 q=x,y,z,w
        q = quaternion_from_euler(math.radians(180), 0, 0)
 
        # 旋转部分进行赋值
        transform.transform.rotation.x = q[0]
        transform.transform.rotation.y = q[1]
        transform.transform.rotation.z = q[2]
        transform.transform.rotation.w = q[3]
 
        # 静态坐标关系发布
        self.static_broadcaster_.sendTransform(transform)
        self.get_logger().info(f'发布 TF: {transform}')

def main():
    rclpy.init()
    node = StaticTFBroadcaster()
    rclpy.spin(node)
    rclpy.shutdown()

File: ROS2-basic/chapter5/chapter5_ws/src/demo_cpp_tf/src/static_tf_broadcaster.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2/LinearMath/Quaternion.hpp"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
#include "tf2_ros/static_transform_broadcaster.h"
#include <memory>


class StaticTFBroadcaster : public rclcpp::Node
{
	public:
		StaticTFBroadcaster() : Node("static_tf_broadcaster")
		{
			broadcaster_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);
			this->publish_tf();
		}
		void publish_tf()
		{
			geometry_msgs::msg::TransformStamped transform;
			transform.header.stamp = this->get_clock()->now();
			transform.header.frame_id = "map";
			transform.child_frame_id = "target_point";
			transform.transform.translation.x = 5.0;
			transform.transform.translation.y = 3.0;
			transform.transform.translation.z = 0.0;
			tf2::Quaternion quat;
			quat.setRPY(0, 0, 60 * M_PI / 180);              // 弧度制欧拉角转四元数
			transform.transform.rotation = tf2::toMsg(quat); // 四元数转ROS消息类型
			broadcaster_->sendTransform(transform);
		}

	private:
		std::shared_ptr<tf2_ros::StaticTransformBroadcaster> broadcaster_;
};

int main(int argc, char **argv)
{
	rclcpp::init(argc, argv);
	rclcpp::spin(std::make_shared<StaticTFBroadcaster>());
	rclcpp::shutdown();
	return 0;
}
